swarm/bmt/bmt.go

```diff
@@ -405,106 +437,124 @@ func (h *Hasher) ReadFrom(r io.Reader) (m int64, err error) {
 // Reset needs to be called before writing to the hasher
 func (h *Hasher) Reset() {
 	h.getTree()
-	h.blockLength = nil
 }
 
 // Hasher implements the SwarmHash interface
 
 // ResetWithLength needs to be called before writing to the hasher
 // the argument is supposed to be the byte slice binary representation of
-// the length of the data subsumed under the hash
-func (h *Hasher) ResetWithLength(l []byte) {
+// the length of the data subsumed under the hash, i.e., span
+func (h *Hasher) ResetWithLength(span []byte) {
 	h.Reset()
-	h.blockLength = l
+	h.bmt.span = span
 }
 
-// Release gives back the Tree to the pool whereby it unlocks
+// releaseTree gives back the Tree to the pool whereby it unlocks
 // it resets tree, segment and index
 func (h *Hasher) releaseTree() {
-	if h.bmt != nil {
-		n := h.bmt.leaves[h.cur]
-		for ; n != nil; n = n.parent {
-			n.unbalanced = false
-			if n.parent != nil {
-				n.root = false
-			}
-		}
-		h.pool.Release(h.bmt)
+	t := h.bmt
+	if t != nil {
+		t.cur = 0
+		t.offset = 0
+		t.span = nil
+		t.hash = nil
 		h.bmt = nil
-
+		h.pool.release(t)
 	}
-	h.cur = 0
-	h.segment = nil
 }
 
-func (h *Hasher) writeSegment(i int, s []byte, d int) {
-	hash := h.pool.hasher()
-	n := h.bmt.leaves[i]
-
-	if len(s) > h.size && n.parent != nil {
-		go func() {
-			hash.Reset()
-			hash.Write(s)
-			s = hash.Sum(nil)
+// writeSegment writes the ith segment into the BMT tree and
+func (h *Hasher) writeSegment(i int, s []byte) {
```

Comment:
Running `go test` with coverage says that `writeSegment` is not covered at all. Could we add a test case that hits this method?

Reply:
Note: it is also not called if you use the public api of `Hasher`, because `sum` is only called in `Sum` with hardcoded `section` parameter value `true`, and with section being true `writeSection` is called instead of `writeSegment`.

Reply:
ok shall i remove it until SegmentWriter is written and has test?

Reply:
Or add a comment that this function is unused, no tests, just created for future use.